### ISSUE TITLE
Added Documentation

### ISSUE THREAD

**Issue by cgoldammer** [JuliaStats/DataFrames.jl#775]
This is an attempt to organize and prioritize added documentation that would be helpful. The priorities allow me to get the most useful content done first. I am planning to get a good part of this done, but likely not all of it.

The list is based on my incomplete information. It describes purely additions to the current documentation. Please get in your feedback - should it be structured differently? Did I forget something important? Did I order the importance incorrectly?

Note 1: I'm planning to update this list continuously based on your comments. Thus, when you add a new comment, you can always see the most up-to-date information.
Note 2: I use bold question marks to ask for your opinion when I'm not sure whether a section should be included or what it should contain.

The following is ordered by importance:
- [ ] Getting started
  - [ ] Adding REPL output to block code when helpful
- [ ] Importing and exporting
  - [ ] RDA
  - [ ] A list to packages that read data of various format (Matlab, SAS, Stata, Excel...)
- [ ] API reference, automatically generated from docstrings (Example: http://pandas.pydata.org/pandas-docs/dev/api.html)
- [ ] Comparison to R & Pandas (Example: http://pandas.pydata.org/pandas-docs/stable/comparison_with_r.html)
- [ ] Extended example: Start from raw data, clean it, create a model, predict, graph results. Use `GadFly` to create graphs.
- [ ] FAQ
  - [ ] Why does this algorithm require a matrix and does not work with a DataFrame?
  - [ ] Why do I need to use symbols for columns instead of strings?
  - [ ] Some of John's blog posts would fit well here (probably shortened)
  - [ ] ...
- [ ] Cookbook (Example: http://pandas.pydata.org/pandas-docs/stable/cookbook.html) **?**
  - [ ] ... various examples to later be filled in
- [ ] Various (not yet organized into sections)
  - [ ] The ways you can generate a dataframe from data
  - [ ] Deleting/appending dataframe or columns
  - [ ] Performance discussion **?**

**Comment by tshort** [comment]
Great outline. As far as a reference API, it'd be nice to use docstrings in the code. These can be used in Julia 0.4 or lower versions with Docile/Lexicon. That way, functions have help in the REPL. I started adding docstrings to some of the code:

https://github.com/JuliaStats/DataFrames.jl/blob/master/src/abstractdataframe/reshape.jl

These docstrings are in Markdown format, so using them with Sphinx will require a conversion with Pandoc or something else. The Markdown docstrings work great with [Mkdocs](http://www.mkdocs.org/), but then the rest of the docs would need to be in Markdown format. Readthedocs supports both Sphinx and Mkdocs, but I don't know if you can mix them.

**Comment by cgoldammer** [comment]
@tshort Yes, for the API reference, I realize using docstrings is the only practical way (this is also how Pandas does it (http://pandas.pydata.org/pandas-docs/dev/api.html). I can look into whether these can be automatically converted from Markdown to Sphinx.

From what I can see, docstrings have relatively small coverage of the code. Is there general agreement to move towards them?

**Comment by tshort** [comment]
Within the Julia community, I think there's a general approval to move towards docstrings. Other DataFrames developers will need to chime in on suitability for DataFrames in particular. I know @johnmyleswhite has used them quite well. We should also think about coordination with DataArrays; there are docstring-like comments in DataArrays, but I'm not sure what the target for those is.

I can help with docstrings for parts of DataFrames that I'm familiar with.

**Comment by cgoldammer** [comment]
I'd be happy to participate in that rewrite. Having recently switched to docstrings in another project, it turned out to be much less work than expected.

**Comment by tshort** [comment]
It looks like the existing documentation uses fairly plain reStructuredText. I just tried converting a couple of the rst files to Markdown with pandoc, and it works well. It might be pretty easy to convert everything to Markdown and then use Mkdocs for the core documentation and the API. I like that approach. Here are examples of packages that do that:

* https://michaelhatherly.github.io/Docile.jl/
* https://tshort.github.io/Sims.jl/
* https://www.google.com/search?q=mkdocs.yml+jl+site:github.com&ie=utf-8&oe=utf-8

@garborg will need to chime in before doing anything drastic like that.

**Comment by garborg** [comment]
Nothing against docstrings and Markdown here, either. I would caution against sinking too much time into DataArrays, though -- removing them completely in favor of something with the less troublesome semantics of the NullableArray concept is a high priority -- hopefully it happens soon.

**Comment by tshort** [comment]
I did a quick-and-dirty conversion to Markdown for use with Mkdocs. Here is the branch:

https://github.com/JuliaStats/DataFrames.jl/tree/mkdocs

Here is the result:

http://juliastats.github.io/DataFrames.jl/

It needs some work, but it might be a starting point if we want to go this route. The API part is quite sparse because we don't have many docstrings, yet.

**Comment by cgoldammer** [comment]
@tshort Thanks for starting on this! My feeling is that this should already be its own issue: "Convert documentation to docstrings" where this can be discussed further? Since conversion is easy, I can start extending the non-API documentation and convert to the format we'll end up choosing in the end.

**Comment by tshort** [comment]
I added some more docstrings to the mkdocs branch and updated the docs in the link above. I'll probably do a WIP pull request on this branch soon.

**Comment by cgoldammer** [comment]
@tshort this is fantastic. How do you want to proceed if I write added documentation? I am thinking of branching off ``mkdocs``, adding some documentation, getting everyone's feedback, and then merging into ``mkdocs`` when it looks good. Let me know if there's a better way to get this done.

**Comment by tshort** [comment]
For proceeding, options could include:

(1) Get what we have merged (after suitable review), and keep submitting PR's for both API and non-API docs as they are done. This assumes we are planning to go with the Mkdocs approach and standardize on Markdown.

(2) Split up the existing Mkdocs branch into a docstrings branch and a non-API docs branch. The non-API docs branch could be either Mkdocs or Sphinx, depending on preference.

Option 1 is the easiest. However, @garborg or other key developers might not be comfortable with the rather sudden switch in documentation formats. Option 1 is not a no-return path. We can keep both the Sphinx and Mkdocs in parallel for a while in case we decide that Mkdocs isn't working. But, any additions on the Markdown side to the non-API docs would have to be ported if we wanted to switch back to Sphinx.

@cgoldammer, if you want to start adding docs immediately, your plan to branch mkdocs sounds fine.

**Comment by johnmyleswhite** [comment]
I think the direction you're headed is the right one.

**Comment by tgg-c-goldammer** [comment]
@tshort and @johnmyleswhite this all sounds good. I have already started writing added documentation in markdown, but I'll wait for feedback from @garborg before putting in pull requests. Then, I'll submit them in the chosen format, one section at a time. I'd also be happy to convert the format afterwards if that becomes necessary.

**Comment by garborg** [comment]
@cgoldammer I'm on board, too -- go for it!

**Comment by cgoldammer** [comment]
@garborg @tshort. Apologies for removing the comment, used the wrong account.

I have written additional documentation in markdown, and I proposed adding it as pull requests one section at a time. So all is good from my side, you'll hear again from me once I have the stuff written.

**Comment by cgoldammer** [comment]
@garborg @tshort. Apologies for removing the comment, used the wrong account.

I have written additional documentation in markdown, and I proposed adding it as pull requests one section at a time. So all is good from my side, you'll hear again from me once I have the stuff written.

**Comment by tshort** [comment]
ReadTheDocs (still with Mkdocs) has a couple of advantages over using Github Pages for hosting:

* It has searching.
* It can host multiple versions.
* The theme will match Julia base as well as other packages.

Here's a test of it:

http://testdataframesjl.readthedocs.org/en/readthedocs/

I like the looks a bit better than the default Mkdocs look on gh-pages.

**Comment by cgoldammer** [comment]
@tshort, I'm going through the `docs` folder and cleaning it up right now - that'll be my first pull request. Do you have preferences for writing section headers? The docs currently use `===` and `---`; I prefer the `#` notation (easier to write and seems cleaner to me) but am happy to stick with the status quo.

**Comment by tshort** [comment]
Great! I prefer `#` notation, too. The `===` just came in as part of the conversion from RST.

**Comment by cgoldammer** [comment]
@tshort, one more question: For code blocks, I assume we want to use \`\`\`julia\`\`\`? It's a little more work but simply seems like the right thing to me.

**Comment by tshort** [comment]
Yes, we want ````julia`. That gives us decent syntax highlighting. The indented blocks came in again as part of the conversion.

**Comment by cgoldammer** [comment]
@tshort Finally done with the cleaning! Since you already merged ```mkdocs``` into ```master```, are you ok with a pull request into ```master```?

**Comment by tshort** [comment]
Yes on the PR to master.

**Comment by cgoldammer** [comment]
I've got additional documentation written and would appreciate your feedback. @johnmyleswhite or @tshort, a quick question about organizing the commits: I'm currently working in my own branch, `added-documentation`. Do you want me to push this new branch to `origin` or instead directly merge this branch into `master`? Thanks!

**Comment by tshort** [comment]
I think pushing your documentation branch here is fine, but @garborg is the main gatekeeper here. If it's here, it's easier for others to push to.

**Comment by garborg** [comment]
@cgoldammer agreed with Tom -- go ahead and push it here.

**Comment by quinnj** [comment]
Closing due to inactivity.
